File: src/ekore/anomalous_dimensions/unpolarized/space_like/aem1.py
"""The :math:`O(a_{em}^1)` Altarelli-Parisi splitting kernels."""

import numba as nb
import numpy as np

from eko import constants

from . import as1


@nb.njit(cache=True)
def gamma_phq(N):
    r"""Compute the leading-order photon-quark anomalous dimension.

    Implements Eq. (2.5) of :cite:`Carrazza:2015dea`.

    Parameters
    ----------
    N : complex
        Mellin moment

    Returns
    -------
    complex
        Leading-order photon-quark anomalous dimension :math:`\\gamma_{\\gamma q}^{(0,1)}(N)`
    """
    return as1.gamma_gq(N) / constants.CF


@nb.njit(cache=True)
def gamma_qph(N, nf):
    r"""Compute the leading-order quark-photon anomalous dimension.

    Implements Eq. (2.5) of :cite:`Carrazza:2015dea`.
    But adding the :math:`N_C` and the :math:`2n_f` factors from :math:`\\theta` inside the
    definition of :math:`\\gamma_{q \\gamma}^{(0)}(N)`.

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors

    Returns
    -------
    complex
        Leading-order quark-photon anomalous dimension :math:`\\gamma_{q \\gamma}^{(0,1)}(N)`
    """
    return as1.gamma_qg(N, nf) / constants.TR * constants.NC


@nb.njit(cache=True)
def gamma_phph(nf):
    r"""Compute the leading-order photon-photon anomalous dimension.

    Implements Eq. (2.5) of :cite:`Carrazza:2015dea`.

    Parameters
    ----------
    nf : int
        Number of active flavors

    Returns
    -------
    complex
        Leading-order phton-photon anomalous dimension :math:`\\gamma_{\\gamma \\gamma}^{(0,1)}(N)`
    """
    nu = constants.uplike_flavors(nf)
    nd = nf - nu
    return 4.0 / 3.0 * constants.NC * (nu * constants.eu2 + nd * constants.ed2)


@nb.njit(cache=True)
def gamma_ns(N, cache):
    r"""Compute the leading-order non-singlet QED anomalous dimension.

    Implements Eq. (2.5) of :cite:`Carrazza:2015dea`.

    Parameters
    ----------
    N : complex
        Mellin moment
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    complex
        Leading-order non-singlet QED anomalous dimension :math:`\\gamma_{ns}^{(0,1)}(N)`
    """
    return as1.gamma_ns(N, cache) / constants.CF


@nb.njit(cache=True)
def gamma_singlet(N, nf, cache):
    r"""Compute the QED leading-order singlet anomalous dimension matrix.

    .. math::
        \\gamma_S^{(0)} = \\left(\begin{array}{cc}
        0 & 0 & 0 & 0 \\
        0 & \\gamma_{\\gamma \\gamma}^{(0,1)} & \\langle e^2 \rangle \\gamma_{\\gamma q}^{(0,1)} & \nu_u e^2_- \\gamma_{\\gamma q}^{(0,1)}\\
        0 & \\langle e^2 \rangle\\gamma_{q \\gamma}^{(0,1)} & \\langle e^2 \rangle \\gamma_{ns}^{(0,1)} & \nu_u e^2_- \\gamma_{ns}^{(0,1)}\\
        0 & \nu_d e^2_- \\gamma_{q \\gamma}^{(0,1)} & \nu_d e^2_- \\gamma_{ns}^{(0,1)} & e^2_\\Delta \\gamma_{ns}^{(0,1)}
        \\end{array}\right)

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    numpy.ndarray
        Leading-order singlet anomalous dimension matrix :math:`\\gamma_{S}^{(0)}(N)`
    """
    e2avg, vue2m, vde2m, e2delta = constants.charge_combinations(nf)
    gamma_ph_q = gamma_phq(N)
    gamma_q_ph = gamma_qph(N, nf)
    gamma_nonsinglet = gamma_ns(N, cache)
    gamma_S_01 = np.array(
        [
            [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
            [
                0.0 + 0.0j,
                gamma_phph(nf),
                e2avg * gamma_ph_q,
                vue2m * gamma_ph_q,
            ],
            [
                0.0 + 0.0j,
                e2avg * gamma_q_ph,
                e2avg * gamma_nonsinglet,
                vue2m * gamma_nonsinglet,
            ],
            [
                0.0 + 0.0j,
                vde2m * gamma_q_ph,
                vde2m * gamma_nonsinglet,
                e2delta * gamma_nonsinglet,
            ],
        ],
        np.complex128,
    )
    return gamma_S_01


@nb.njit(cache=True)
def gamma_valence(N, nf, cache):
    r"""Compute the QED leading-order valence anomalous dimension matrix.

    .. math::
        \\gamma_V^{(0,1)} = \\left(\begin{array}{cc}
        \\langle e^2 \rangle \\gamma_{ns}^{(0,1)} & \nu_u e^2_- \\gamma_{ns}^{(0,1)}\\
        \nu_d e^2_- \\gamma_{ns}^{(0,1)} & e^2_\\Delta \\gamma_{ns}^{(0,1)}
        \\end{array}\right)

    Parameters
    ----------
    N : complex
        Mellin moment
    nf : int
        Number of active flavors
    cache: numpy.ndarray
        Harmonic sum cache

    Returns
    -------
    numpy.ndarray
        Leading-order singlet anomalous dimension matrix :math:`\\gamma_{S}^{(0)}(N)`
    """
    e2avg, vue2m, vde2m, e2delta = constants.charge_combinations(nf)
    gamma_V_01 = np.array(
        [
            [e2avg, vue2m],
            [vde2m, e2delta],
        ],
        np.complex128,
    )
    return gamma_V_01 * gamma_ns(N, cache)

File: crates/ekore/src/anomalous_dimensions/unpolarized/spacelike/aem1.rs
//! |LO| |QED|.
use num::complex::Complex;
use num::Zero;

use crate::cmplx;
use crate::constants::{ChargeCombinations, CF, ED2, EU2, NC, TR};
use crate::harmonics::cache::Cache;

use crate::anomalous_dimensions::unpolarized::spacelike::as1;

/// Compute the photon-quark anomalous dimension.
///
/// Implements Eq. (2.5) of [\[Carrazza:2015dea\]][crate::bib::Carrazza2015dea].
pub fn gamma_phq(c: &mut Cache, nf: u8) -> Complex<f64> {
    as1::gamma_gq(c, nf) / CF
}

/// Compute the quark-photon anomalous dimension.
///
/// Implements Eq. (2.5) of [\[Carrazza:2015dea\]][crate::bib::Carrazza2015dea].
/// However, we are adding the $N_C$ and the $2n_f$ factors from $\theta$ inside the
/// definition of $\gamma_{q \gamma}^{(0)}(N)$.
pub fn gamma_qph(c: &mut Cache, nf: u8) -> Complex<f64> {
    as1::gamma_qg(c, nf) / TR * (NC as f64)
}

/// Compute the photon-photon anomalous dimension.
///
/// Implements Eq. (2.5) of [\[Carrazza:2015dea\]][crate::bib::Carrazza2015dea].
pub fn gamma_phph(_c: &mut Cache, nf: u8) -> Complex<f64> {
    let cc = ChargeCombinations { nf };
    cmplx!(
        4.0 / 3.0 * (NC as f64) * ((cc.nu() as f64) * EU2 + (cc.nd() as f64) * ED2),
        0.
    )
}

/// Compute the non-singlet anomalous dimension.
///
/// Implements Eq. (2.5) of [\[Carrazza:2015dea\]][crate::bib::Carrazza2015dea].
pub fn gamma_ns(c: &mut Cache, nf: u8) -> Complex<f64> {
    as1::gamma_ns(c, nf) / CF
}

/// Compute the singlet anomalous dimension matrix.
pub fn gamma_singlet(c: &mut Cache, nf: u8) -> [[Complex<f64>; 4]; 4] {
    let cc = ChargeCombinations { nf };

    let gamma_ph_q = gamma_phq(c, nf);
    let gamma_q_ph = gamma_qph(c, nf);
    let gamma_nonsinglet = gamma_ns(c, nf);

    [
        [
            Complex::<f64>::zero(),
            Complex::<f64>::zero(),
            Complex::<f64>::zero(),
            Complex::<f64>::zero(),
        ],
        [
            Complex::<f64>::zero(),
            gamma_phph(c, nf),
            cc.e2avg() * gamma_ph_q,
            cc.vue2m() * gamma_ph_q,
        ],
        [
            Complex::<f64>::zero(),
            cc.e2avg() * gamma_q_ph,
            cc.e2avg() * gamma_nonsinglet,
            cc.vue2m() * gamma_nonsinglet,
        ],
        [
            Complex::<f64>::zero(),
            cc.vde2m() * gamma_q_ph,
            cc.vde2m() * gamma_nonsinglet,
            cc.e2delta() * gamma_nonsinglet,
        ],
    ]
}

/// Compute the valence anomalous dimension matrix.
pub fn gamma_valence(c: &mut Cache, nf: u8) -> [[Complex<f64>; 2]; 2] {
    let cc = ChargeCombinations { nf };
    let g = gamma_ns(c, nf);
    [
        [cc.e2avg() * g, cc.vue2m() * g],
        [cc.vde2m() * g, cc.e2delta() * g],
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{assert_approx_eq_cmplx, cmplx};
    use num::complex::Complex;
    use num::Zero;
    const NF: u8 = 5;

    #[test]
    fn number_conservation() {
        const N: Complex<f64> = cmplx!(1., 0.);
        let mut c = Cache::new(N);
        let me = gamma_ns(&mut c, NF);
        assert_approx_eq_cmplx!(f64, me, Complex::zero(), epsilon = 1e-12);
    }

    #[test]
    fn quark_momentum_conservation() {
        const N: Complex<f64> = cmplx!(2., 0.);
        let mut c = Cache::new(N);
        let me = gamma_ns(&mut c, NF) + gamma_phq(&mut c, NF);
        assert_approx_eq_cmplx!(f64, me, Complex::zero(), epsilon = 1e-12);
    }

    #[test]
    fn photon_momentum_conservation() {
        const N: Complex<f64> = cmplx!(2., 0.);
        let mut c = Cache::new(N);

        for nf in 2..7 {
            let cc = ChargeCombinations { nf };
            assert_approx_eq_cmplx!(
                f64,
                EU2 * gamma_qph(&mut c, cc.nu())
                    + ED2 * gamma_qph(&mut c, cc.nd())
                    + gamma_phph(&mut c, nf),
                cmplx!(0., 0.),
                epsilon = 2e-6
            );
        }
    }
}
